HW0/P4/P4.py

```diff
@@ -40,37 +47,103 @@
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    # Read P4_measurements.txt file and extract coordinates from the original arrays
+    # for ploting measurements
+
+    measurements_array = np.loadtxt('P4_measurements.txt', delimiter = ',')
+    fix_stretching = np.diag((1/float(rx), 1/float(ry), 1/float(rz)))
+    approx_postions = np.dot(fix_stretching, measurements_array.T)
+    x_coords = approx_postions[0,:]
+    y_coords = approx_postions[1,:]
+    z_coords = approx_postions[2,:]
+    ax.plot(x_coords, y_coords, z_coords,'.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    #Initialize propagation matrix A
+    A = np.zeros((6,6))
+    (A[0,0],A[1,1],A[2,2]) = [1] * 3
+    (A[0,3],A[1,4],A[2,5]) = [dt] * 3
+    (A[3,3],A[4,4],A[5,5]) = [1-c * dt] * 3
+
+    #Initialize vector a
+    a = np.array([0,0,0,0,0,g*dt]).T
 
-    # Initial conditions for s0
-    # Compute the rest of sk using Eq (1)
+    #Initialize a 6 * K matrix s to store all states
+    s = np.zeros((6, K))
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    #Assign s0 as the initial state
+    s[:,0] = np.array([0,0,2,15,3.5,4.0]).T
+
+    #Compute all of the sk for k = 0, 1, . . . , K - 1 using Eq (1).
+    for i in range(1, K):
+        s[:, i] = np.dot(A, s[:, (i - 1)]) + a
+
+    #Convert it to arrays before plotting the first three rows
+    x_coords = s[0, :]
+    y_coords = s[1, :]
+    z_coords = s[2, :]
+
+    # Plot the predicted positions, xk , using the commented ax.plot command.
+    ax.plot(x_coords, y_coords, z_coords,'-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    #Initialize matrix B and C
+    B = np.diag((bx, by, bz, bvx, bvy, bvz))
+    C = np.hstack([np.diag((rx,ry, rz)), np.zeros((3,3))])
+
+    #Define four python functions for prediction steps and update steps
+
+    #predictS: Implements equation (2) and returns predicted steps
+    def predictS(sk):
+        si = np.dot(A, sk) + a
+        return(si)
+
+    #predictSig: Implements equation (3) and returns predicted sigma.
+    def predictSig(sigma):
+        p_sig = np.dot(np.dot(A, sigma), A.T) + np.dot(B, B.T)
+        p_sig = np.asmatrix(p_sig).I
+        return(p_sig)
+
+    #updateSig: Implements equation (3) and returns updated sigma.
+    def updateSig(p_sig):
+        sig = p_sig + np.dot(C.T, C)
+        sig = np.asmatrix(sig).I
+        return(sig)
+
+    #updateSig: Implements equation (4) and returns updated step.
+    def updateS(si, p_sig, sig, m):
+        sk = np.dot(sig, (np.dot(p_sig, si) + np.dot(C.T, m)).T)
+        return sk
 
```

Comment:
You could have added the matrix A, C and B as input parameters in the above functions, in case you wanted to simulate different data without having to update the definitions of the functions manually ! It wasn't asked in the question but it is something I often do :-)